s3cmd

```diff
@@ -583,12 +594,79 @@ def cmd_object_del(args):
 
         if not recursive:
             rc = subcmd_object_del_uri(uri_str)
-        else:
+        elif Config().exclude:
             rc = subcmd_batch_del(uri_str = uri_str)
+        else:
+            rc = subcmd_batch_del_iterative(uri_str = uri_str)
         if not rc:
             return rc
     return EX_OK
 
+def subcmd_batch_del_iterative(uri_str = None, bucket = None):
+    """ Streaming version of batch deletion (doesn't realize whole list in memory before deleting). 
+
+    Differences from subcmd_batch_del:
+      - Obeys cfg.max_delete, but if it is >1000, will delete files until we are sure that at least 
+        that many files are to be deleted
+      - Only prints out individual file names if <1000 files are to be deleted
+      - Does not obey --exclude directives (use subcmd_batch_del in those cases)
+
+    """
+    if bucket and uri_str:
+        raise ValueError("Pass only one of uri_str or bucket")
+    if bucket: # bucket specified
+        uri_str = "s3://%s" % bucket
+    s3 = S3(cfg)
+    uri = S3Uri(uri_str)
+    bucket = uri.bucket()
+
+    to_delete = []
+    deleted = [0]
+    deleted_bytes = [0]
+
+    def _status():
+        total_size, size_coeff = formatSize(deleted_bytes[0], Config().human_readable_sizes)
+        total_size_str = str(total_size) + size_coeff
+        output(u"Deleted %s -> %s from %s" % (str(deleted[0]).ljust(8), total_size_str.ljust(8), uri))
```

Comment:
1) Should be a "debug" message as for batch_del.
2) The formatting is very strange, if i'm not wrong it will looks like that:
"Deleted 15 -> 3Mb from s3://mybucket/..."
??? total -> size ???